pkg/plugin/datasource.go

```diff
@@ -0,0 +1,326 @@
+	"github.com/grafana/grafana-plugin-sdk-go/backend/resource/httpadapter"
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+// Ensure it actually implements the interfaces we need it to
+var (
+	_ backend.QueryDataHandler    = (*Datasource)(nil)
+	_ backend.CheckHealthHandler  = (*Datasource)(nil)
+	_ backend.CallResourceHandler = (*Datasource)(nil)
+)
+
+const MAX_RESULTS = 10000 // same as what the actual Cribl UI imposes
+const QUERY_PAGE_SIZE = 1000
+const CRIBL_TIME_FIELD = "_time"
+const MAX_DELAY_WAITING_FOR_FINISHED = 60 * time.Second // could make this configurable in a future release
+const MAX_BACKOFF_DURATION = 2 * time.Second
+const GRAFANA_TIME_FIELD_NAME = "Time"
+
+// Expose a counter metric tracking the # of queries, broken down by type (adhoc vs. savedSearchId)
+var queryCounter = promauto.NewCounterVec(
+	prometheus.CounterOpts{
+		Namespace: "grafana_plugin", // recommended by Grafana in their "Best Practices" section
+		Name:      "cribl_search_queries_total",
+		Help:      "Total number of queries.",
+	},
+	[]string{"query_type"},
+)
+
+// Expose a counter metric tracking the total # of result events produced by queries
+var resultsCounter = promauto.NewCounterVec(
+	prometheus.CounterOpts{
+		Namespace: "grafana_plugin",
+		Name:      "cribl_search_results_total",
+		Help:      "Total number of events in results.",
+	},
+	[]string{"query_type"},
+)
+
+type Datasource struct {
+	ResourceHandler backend.CallResourceHandler
+	Settings        *models.PluginSettings
+	SearchAPI       *SearchAPI
+}
+
+// NewDatasource creates a new datasource instance.
+func NewDatasource(ctx context.Context, settings backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
+	ps, err := models.LoadPluginSettings(settings)
+	if err != nil {
+		return nil, err
+	}
+	ds := &Datasource{}
+	ds.Settings = ps
+	ds.SearchAPI = NewSearchAPI(ps)
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/savedSearchIds", ds.handleSavedSearchIds)
+	ds.ResourceHandler = httpadapter.New(mux)
+
+	return ds, nil
+}
+
+// QueryData handles multiple queries and returns multiple responses.
+// req contains the queries []DataQuery (where each query contains RefID as a unique identifier).
+// The QueryDataResponse contains a map of RefID to the response for each query, and each response
+// contains Frames ([]*Frame).
+func (d *Datasource) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	// create response struct
+	response := backend.NewQueryDataResponse()
+
+	// loop over queries and execute them individually.
+	for _, q := range req.Queries {
+		res := d.query(ctx, req.PluginContext, q)
+
+		// save the response in a hashmap
+		// based on with RefID as identifier
+		response.Responses[q.RefID] = res
+	}
+
+	return response, nil
+}
+
```

Comment:
NIT: Noticed other function names are CamelCase but this one is lower.

Reply:
It's a convention.  PublicMethod vs. privateMethod.